### DISCUSSION TITLE
Prospects for support of numpy.broadcast_arrays

### DISCUSSION THREAD

**Discussion by njlivesey** [hgrecco/pint#1607] [Q&A]
Dear all, (This is my first time posting on github, so apologies for any etiquette fails.)

   Am I right in concluding that pint does not currently support numpy.broadcast_arrays?  What are the prospects that it might do down the road (with subok=True, presumably)?

   I'm asking because I'm exploring porting an existing codebase from astropy.units to pint, and I've used broadcast_arrays in several places (though I may be able to work around).  Apologies in advance if other similar requests emerge that could have been combined.

Nathaniel Livesey

**Comment by hgrecco** [comment]
Welcome and as long as you are nice to others don't worry about etiquette fails.

Can you post the code here, and I will be happy to take a look and give you a more detailed answer

**Comment by andrewgsavage** [comment]
I had a go at implementing this here https://github.com/hgrecco/pint/pull/1610
I'm not familiar with how subok is used so please check it works as you expect.

**Reply by andrewgsavage** [reply]
I tagged this discussion in the PR. Unfortunately it doesn't show the PR here like it does for issues.

**Reply by njlivesey** [reply]
Many thanks indeed.

I'll try to pull this and test it in my setup (will probably have to wait till Friday I'm afraid).  I do note the use of the "all_consistent" flag in the updated code and I'm not sure that it is warranted.  I can imagine many cases where one would want to broadcast quantities of different dimensionality to the same shape (e.g., latitude/longitude/height coordinates of geographical points; temperature, density, and velocities of a fluid in some grid).  Perhaps I'm misunderstanding the intent of the option, and doing that will work fine. Apologies for being a worry-wart if so.

I take it it would have been better to open this as an issue rather than in the "discussions", right?  I can do that in future.  Spoiler alert, I'll be asking about numpy.diag and numpy.fill_diagonal next. (Background, I'm moving a previously developed codebase over from astropy.units and so am noticing the gaps - not that astropy didn't have its own gaps!).

**Reply by andrewgsavage** [reply]
> Many thanks indeed.
> 
> I'll try to pull this and test it in my setup (will probably have to wait till Friday I'm afraid). I do note the use of the "all_consistent" flag in the updated code and I'm not sure that it is warranted. I can imagine many cases where one would want to broadcast quantities of different dimensionality to the same shape (e.g., latitude/longitude/height coordinates of geographical points; temperature, density, and velocities of a fluid in some grid). P

A pint Quantity can only store one dimensionality, so I think this won't be possible

> I take it it would have been better to open this as an issue rather than in the "discussions", right? 

yea there's a button on the side that lets you convert it to an issue too

**Comment by jules-ch** [comment]
Can you test this in the new version ?

**Comment by njlivesey** [comment]
Hi Jules,

  Sorry, probably not for a couple of weeks.  I’m on back-to-back travel this week and for the next two, with next week’s meeting being perhaps one of the most important of my career (whether to continue operating our satellite instrument or not).

Nathaniel


Nathaniel Livesey
Mail Stop 183-701, Jet Propulsion Laboratory
4800 Oak Grove Drive, Pasadena, California 91109.
Phone: +1 818 354-4214   Fax: +1 818 393-5065


From: Jules Chéron ***@***.***>
Date: Thursday, October 27, 2022 at 7:51 AM
To: hgrecco/pint ***@***.***>
Cc: Livesey, Nathaniel J (US 3290) ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [hgrecco/pint] Prospects for support of numpy.broadcast_arrays (Discussion #1607)

Can you test this in the new version ?

**Reply by hgrecco** [reply]
Good luck in your meeting!

**Comment by njlivesey** [comment]
Hi Hernan,

   Many thanks, and do bug me again in a couple of weeks if I appear to have forgotten.  I won’t mind (that’s the kind of fun work I like to do evenings and weekends when I fancy a change from the NASA bureaucracy).

Yours,

Nathaniel

Nathaniel Livesey
Mail Stop 183-701, Jet Propulsion Laboratory
4800 Oak Grove Drive, Pasadena, California 91109.
Phone: +1 818 354-4214   Fax: +1 818 393-5065


From: Hernan Grecco ***@***.***>
Date: Thursday, October 27, 2022 at 5:02 PM
To: hgrecco/pint ***@***.***>
Cc: Livesey, Nathaniel J (US 3290) ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [hgrecco/pint] Prospects for support of numpy.broadcast_arrays (Discussion #1607)

Good luck in your meeting!

**Reply by njlivesey** [reply]
Hi again,

   I found some time today to give it a go. It worked fine for quantities with the same units (Dimensionality).

   However, I'm afraid the "all_consistent" (and probably the "match_input") flavor of the implementation would present a problem for my intended use.  I guess we can debate whether or not it should be OK to broadcast arrays with different units to the same shape as each other.  However, my sense is that to insist they be the same Dimensionality (in a units sense) seems like an unnecessary restriction to me.  Consider the case where we have arrays describing temperature and density of some grid, being able to broadcast these quantities to the same shape would be useful.

Thanks for looking into this.

Nathaniel

**Comment by andrewgsavage** [comment]
can you put that example into code?

**Comment by njlivesey** [comment]
Sure, here goes:

```python
import numpy as np
import pint

ureg = pint.UnitRegistry()

T = np.ones(shape=[10,1]) * 250.0 * ureg.K
v = np.ones(shape=[1,10]) * 10.0 * ureg.m / ureg.s
rho = np.ones(shape=[10,10]) * ureg.kg / ureg.m**3       
        
T, v, rho = np.broadcast_arrays(T, v, rho)
```

Hope this helps.

Nathaniel
